### PR TITLE
Membership form validation

### PR DESCRIPTION
**Pull request by KaiNissen**
I moved the fee validation stuff around to be able to just validate that rather than validate the whole form. Tests still need to be moved.

### CODE REVIEW

**Comment by JeroenDeDauw** [comment]
Is this still WIP? Kinda difficult for me to continue with the validator now

**Comment by KaiNissen** [comment]
Sorry, @JeroenDeDauw. I still wanted to move the stuff that is testing the `MembershipFeeValidator` out of the general `MembershipApplicationValidator`.

**Comment by JeroenDeDauw** [comment]
Not to worry, still had other stuff to poke at. Still, would be nice to get this into a state that can be merged.

File: app/js/tests/test_form_validation.js [hunk 1]
```diff
@@ -71,7 +71,8 @@ test( 'Given a private adddress, address validation sends values to server', fun
 } );
 
 test( 'Given an incomplete private adddress, address validation sends no values to server', function ( t ) {
-	var postFunctionSpy = sinon.spy(),
+	var negativeResult = { status: 'ERR' },
```
Comment (JeroenDeDauw) [inline app/js/tests/test_form_validation.js]: failureResult probably better

File: src/UseCases/ApplyForMembership/ApplicationValidationResult.php [hunk 2]
```diff
@@ -24,6 +24,14 @@ public function __construct( array $violations = [] ) {
 		$this->violations = $violations;
 	}
 
+	public function addViolations( array $violations ) {
+		$this->violations = array_merge( $this->violations, $violations );
+	}
```
Comment (JeroenDeDauw) [inline src/UseCases/ApplyForMembership/ApplicationValidationResult.php]: Why is this mutability needed?

File: tests/Unit/UseCases/ApplyForMembership/MembershipApplicationValidatorTest.php [hunk 3]
```diff
@@ -2,103 +2,71 @@
...
 use WMDE\Fundraising\Frontend\Tests\Data\ValidMembershipApplicationRequest;
 use WMDE\Fundraising\Frontend\UseCases\ApplyForMembership\ApplicationValidationResult as Result;
+use WMDE\Fundraising\Frontend\UseCases\ApplyForMembership\ApplicationValidationResult;
 use WMDE\Fundraising\Frontend\UseCases\ApplyForMembership\ApplyForMembershipRequest;
 use WMDE\Fundraising\Frontend\UseCases\ApplyForMembership\MembershipApplicationValidator;
+use WMDE\Fundraising\Frontend\Validation\MembershipFeeValidator;
 
 /**
- * @covers WMDE\Fundraising\Frontend\UseCases\ApplyForMembership\MembershipApplicationValidator
+ * @covers WMDE\Fundraising\Frontend\Validation\MembershipFeeValidator
```
Comment (JeroenDeDauw) [inline tests/Unit/UseCases/ApplyForMembership/MembershipApplicationValidatorTest.php]: huh?

File: tests/Unit/UseCases/ApplyForMembership/MembershipApplicationValidatorTest.php [hunk 4]
```diff
@@ -2,103 +2,71 @@
 
 declare( strict_types = 1 );
 
-namespace WMDE\Fundraising\Frontend\Tests\Unit\UseCases\ApplyForMembership;
+namespace WMDE\Fundraising\Frontend\Tests\Unit\Validation;
```
Comment (JeroenDeDauw) [inline tests/Unit/UseCases/ApplyForMembership/MembershipApplicationValidatorTest.php]: This is not where this file is

**Comment by JeroenDeDauw** [comment]
Looks good apart from some small things to me, though @gbirke should check the JS as well

File: app/js/lib/form_validation.js [hunk 5]
```diff
@@ -24,9 +24,9 @@ var jQuery = require( 'jquery' ),
 					return { status: 'OK' };
 			}
 
-			// Skip validation if not all required fields have been filled
+			// Return error status if not all required fields have been filled
 			if ( this.formValuesHaveEmptyRequiredFields( formValues, requiredFields ) ) {
-				return null;
+				return { status: 'ERR' };
```
Comment (gbirke) [inline app/js/lib/form_validation.js]: The tri-state return (null, object with status ERR and object with status OK) was deliberate (see https://github.com/wmde/FundraisingFrontend/blob/master/app/js/lib/redux_validation.js#L41-L43 ). 
Let's discuss the different approaches and their pros and cons in person.

File: src/UseCases/ApplyForMembership/ApplicationValidationResult.php (hunk 2 above)
Comment (KaiNissen) [inline src/UseCases/ApplyForMembership/ApplicationValidationResult.php]: It's not. I changed it.

**Comment by KaiNissen** [comment]
Sorry, I forced pushed my changes and overwrote @gbirke's changes. Cherry-picked the commit, though.

**Pull request merged**